os/drivers/video/video_null_driver.c

```diff
@@ -0,0 +1,1554 @@
+	}
+
+	for (x = 0, y = 1, i = n; i > 0; --i) {
+		r = y;
+		y = arbt[i - 1] * y + x;
+		x = r;
+	}
+
+	*ntor = y;
+	*dtor = x;
+}
+
+/****************************************************************************
+ * Name: video_null_convert_interval
+ *
+ * Description:
+ *   This is a helper function to calulate frame interval from it numerator &
+ *   denominator.
+ *
+ * Input Parameters:
+ *   ntor - frame numerator.
+ *   dtor - frame denominator
+ *
+ * Returned Values:
+ *   Return interval value.
+ *
+ ****************************************************************************/
+static uint32_t video_null_convert_interval(uint32_t ntor, uint32_t dtor)
+{
+	uint32_t multi = 10000000;
+
+	if (dtor == 0 || ntor / dtor >= ((uint32_t) - 1) / multi) {
+		return (uint32_t) - 1;
+	}
+
+	while (ntor > ((uint32_t) - 1) / multi) {
+		multi /= 2;
+		dtor /= 2;
+	}
+
+	return dtor ? ntor * multi / dtor : 0;
+}
+
+/****************************************************************************
+ * Name: video_null_try_frame_interval
+ *
+ * Description:
+ *   This function is call to check the frame interval can be applied to device.
+ *
+ * Input Parameters:
+ *   frame - Pointer to frame interval.
+ *   interval - Interval to to match with frame.
+ *
+ * Returned Values:
+ *   OK on success, appropriate error code on failures
+ *
+ ****************************************************************************/
+static uint32_t video_null_try_frame_interval(struct v4l2_frames_s *frame, uint32_t interval)
+{
+	int i = 0;
+	uint32_t min;
+	uint32_t max;
+	uint32_t step;
+	uint32_t val;
+
+	if (frame->type == V4L2_FRMIVAL_TYPE_DISCRETE) {
+		uint32_t best = -1, dist;
+		for (i = 0; i < frame->intervalcount; ++i) {
+			dist = interval > frame->intervals[i].min_numerator ? interval - frame->intervals[i].min_numerator : frame->intervals[i].min_numerator - interval;
+
+			if (dist > best) {
+				break;
+			}
+
+			best = dist;
+		}
+		interval = frame->intervals[i - 1].min_numerator;
+	} else if (frame->type == V4L2_FRMIVAL_TYPE_STEPWISE) {
+		min = frame->intervals[0].min_numerator;
+		max = frame->intervals[0].max_numerator;
+		step = frame->intervals[0].step_numerator;
+
+		interval = min + (interval - min + step / 2) / step * step;
+		if (interval > max) {
+			interval = max;
+		}
+	}
+
+	return interval;
+}
+
+/****************************************************************************
+ * Name: video_null_init_transfer
+ *
+ * Description:
+ *   This function open respective frame file and fill the data to transfer
+ *   buffer.
+ *
+ * Input Parameters:
+ *   priv - Pointer to device object.
+ *
+ * Returned Value:
+ *   0 on success, respective error code on failure.
+ *
+ ****************************************************************************/
+int video_null_init_transfer(FAR null_priv_t *priv)
+{
+	int ret = OK;
+	int fd;
+	char filename[FILE_NAMELEN] = { 0, };
+
+	if (!priv->cur_format) {
+		return -EINVAL;
+	}
+
+	if (!priv->cur_frame) {
```

Comment:
There's no log, so (A == B || C == D) will be better

Reply:
Done